sql/src/main/java/org/apache/druid/sql/calcite/schema/MetadataSegmentView.java

```diff
@@ -0,0 +1,255 @@
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStart;
+import org.apache.druid.java.util.common.lifecycle.LifecycleStop;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.java.util.http.client.Request;
+import org.apache.druid.server.coordinator.BytesAccumulatingResponseHandler;
+import org.apache.druid.sql.calcite.planner.PlannerConfig;
+import org.apache.druid.timeline.DataSegment;
+import org.jboss.netty.handler.codec.http.HttpMethod;
+import org.joda.time.DateTime;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * This class polls the coordinator in background to keep the latest published segments.
+ * Provides {@link #getPublishedSegments()} for others to get segments in metadata store.
+ */
+@ManageLifecycle
+public class MetadataSegmentView
+{
+
+  private static final EmittingLogger log = new EmittingLogger(MetadataSegmentView.class);
+
+  private final DruidLeaderClient coordinatorDruidLeaderClient;
+  private final ObjectMapper jsonMapper;
+  private final BytesAccumulatingResponseHandler responseHandler;
+  private final BrokerSegmentWatcherConfig segmentWatcherConfig;
+
+  private final boolean isCacheEnabled;
+  @Nullable
+  private final ConcurrentMap<DataSegment, DateTime> publishedSegments;
+  private final ScheduledExecutorService scheduledExec;
+  private final long pollPeriodInMS;
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+  private final AtomicBoolean cachePopulated = new AtomicBoolean(false);
+
+  @Inject
+  public MetadataSegmentView(
+      final @Coordinator DruidLeaderClient druidLeaderClient,
+      final ObjectMapper jsonMapper,
+      final BytesAccumulatingResponseHandler responseHandler,
+      final BrokerSegmentWatcherConfig segmentWatcherConfig,
+      final PlannerConfig plannerConfig
+  )
+  {
+    Preconditions.checkNotNull(plannerConfig, "plannerConfig");
+    this.coordinatorDruidLeaderClient = druidLeaderClient;
+    this.jsonMapper = jsonMapper;
+    this.responseHandler = responseHandler;
+    this.segmentWatcherConfig = segmentWatcherConfig;
+    this.isCacheEnabled = plannerConfig.isMetadataSegmentCacheEnable();
+    this.pollPeriodInMS = plannerConfig.getMetadataSegmentPollPeriod();
+    this.publishedSegments = isCacheEnabled ? new ConcurrentHashMap<>(1000) : null;
+    this.scheduledExec = Execs.scheduledSingleThreaded("MetadataSegmentView-Cache--%d");
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+    try {
+      if (isCacheEnabled) {
```

Comment:
Looks like this class violates the Single Responsibility Principle, see #7392.

Reply:
hmm I need to read up more on Single Responsibility Principle, another way to look at it, is the responsibility of this class is to get metadata segments, the public method is `getPublishedSegments`, then it kind of does one thing i.e. get metadata segments.

Reply:
Probably SRP is a wrong link in this case, cohesion may be a more correct reference. Anyway, it seems to me that `isCacheEnabled` replaces a perfectly healthy polymorphism "CachingMetadataSegmentView"/"DirectMetadataSegmentView" here. Somebody may want to add "CachingWithDirectMetadataStoreAccess" (vs. using Coordinator as a proxy), etc.